src/minisat/mtl/MapWithKeys.h

```diff
@@ -0,0 +1,59 @@
+//
+// Created by Antti on 18.05.21.
+//
+
+#ifndef OPENSMT_KEYVALMAP_H
+#define OPENSMT_KEYVALMAP_H
+#include "Map.h"
+
+template<class K, class D, class H, class E = Equal<K> >
+class MapWithKeys {
+    Map<K,D,H,E> map;
+    vec<K> keys;
+public:
+    int getSize() { return map.getSize(); }
+    MapWithKeys () {}
+    MapWithKeys (const H& h, const E& e) : map(h, e) {}
+    MapWithKeys (MapWithKeys&& o) noexcept : map(std::move(o.map)), keys(std::move(o.keys)) { }
+
+    MapWithKeys&  operator = (MapWithKeys&& o) { std::swap(map, o.map); std::swap(keys, o.keys); return *this; }
+
+    // PRECONDITION: the key must already exist in the map.
+    const D& operator [] (const K& k) const
+    {
+        return map.operator[](k);
+    }
+
+    // PRECONDITION: the key must already exist in the map.
+    D& operator [] (const K& k)
+    {
+        return map.operator[](k);
+    }
+
+    // PRECONDITION: the key must *NOT* exist in the map.
+    void insert (const K& k, const D& d) { map.insert(k, d); keys.push(k); }
```

Comment:
We could add `assert(not has(k))` here to enforce the precondition in Debug.

Reply:
Ok. I'm going to put it in mtl::Map.